llama_index/callbacks/base.py

```diff
@@ -135,29 +135,34 @@ def as_trace(self, trace_id: str) -> Generator[None, None, None]:
 
     def start_trace(self, trace_id: Optional[str] = None) -> None:
         """Run when an overall trace is launched."""
-        if not self._trace_id:
-            self._reset_trace_events()
+        if trace_id is not None:
+            if self._trace_id_stack is None:
+                self._reset_trace_events()
 
-            for handler in self.handlers:
-                handler.start_trace(trace_id=trace_id)
+                for handler in self.handlers:
+                    handler.start_trace(trace_id=trace_id)
 
-            self._trace_id = trace_id
+                self._trace_id_stack = [trace_id]
+            else:
+                self._trace_id_stack.append(trace_id)
 
     def end_trace(
         self,
         trace_id: Optional[str] = None,
         trace_map: Optional[Dict[str, List[str]]] = None,
     ) -> None:
         """Run when an overall trace is exited."""
-        if trace_id is not None and trace_id == self._trace_id:
-            for handler in self.handlers:
-                handler.end_trace(trace_id=trace_id, trace_map=self._trace_map)
-            self._trace_id = None
+        if trace_id is not None and self._trace_id_stack is not None:
+            self._trace_id_stack.pop()
+            if len(self._trace_id_stack) == 0:
+                for handler in self.handlers:
+                    handler.end_trace(trace_id=trace_id, trace_map=self._trace_map)
+                self._trace_id_stack = None

```

Comment:
should we throw an error if trace id stack is None / should that ever happen here?